### PR TITLE
3 inventory and items

### LINKED ISSUE #3: [Team 3] Items/Inventory
# **Description**
This feature introduces a system for players to collect and manage various items throughout the game. The inventory will be available through an in-game menu, where players can view, organize, select and interact with items. It is essential for adding depth to gameplay by providing players with the ability to interact meaningfully with the world by gathering and using resources to enhance gameplay.

## SubFeature/component
- Item Abstract Class (#31)
- Resources/Potions/Artefacts(#24 )
- Inventory Animal & Item Storage
- Inventory Interaction (Interface)
- Inventory UI View
- Scroll 
- Weapons (#30)

## **Examples**
- Pokemon style animal inventory, a team of 6 with no box storage prompting the user to switch one of the monsters for the new one. 
- Bag style of items, can store up to 10 items ensuring that the player uses the items

# **Dependencies**

Animal inventory depends on Animal Player team (#4).

# **Milestones**
- [x] Develop Inventory Structure (Aug. 20)
- [x]  Create Item abstract structure (Aug 20)
- [x] Have basic inventory UI (Aug. 23)
- [x] Create Resource/Potion structure (Aug 23)
- [x] Develop function UI for visualisation of items in inventory (Aug 24)
- [x] Develop function UI for visualisation of items in map (Aug 24)
- [x] Be to pick up/hold items from map into inventory (Aug 25)
- [x] Be able to use items from the inventory (Aug 27)
- [x] Complete Inventory (Aug 27)
- [ ] Merge into Main (Aug 27)

**Completion Deadline:** Aug. 28

# **Documentation**
- [Inventory Wiki Page](https://github.com/UQcsse3200/2024-studio-2/wiki/Inventory)
- [Items Wiki Page](https://github.com/UQcsse3200/2024-studio-2/wiki/Items)

# **Members**
- Senuri Panadura Arachchige (s-shalomi) 
- Thomas Preston (tompreston1)
- Lewis Hickson (lhick2108)
- Mittun Sudhahar (mitsudz)
- Phong Nguyen (beannie)

### LINKED ISSUE #18: Create Item Abstract Class
# **Description**
**Task:** [Create Item Abstract Class](#3).
**Feature:** [#3 Items/Inventory]
 
Initialise the abstract class for items. This class will have functions and fields which can be inherited by other subclasses such as all item subclasses.
 
# **Milestones**
List of steps that need to be completed for this task.
- [x] Add Abstract Class (Aug. 11)
- [x] Add Tests for Abstract Class (Aug. 12)
- [x] Add Documentation to wiki (Aug. 14)
- [ ] Merge into main (27 Aug)

**Completion Deadline:** Aug. 14

# **Documentation**
- [Abstract Item Documentation](https://github.com/UQcsse3200/2024-studio-2/wiki/Abstract-Item)
 
# **Member**
- Name (GitHub username) `Discord Username`
- Mittun (@mitsudz) `mitsudz`

### LINKED ISSUE #19: Inventory Class
# **Description**
**Task:** [Create Inventory Abstract Class](#3 ).
**Feature:** [Items and Inventory](#3 )

A class to support various types of inventory's (e.g. combat inventory, general inventory etc.). Should implement abstract interface for handling using, adding and removing from inventory. 
 
# **Milestones**
List of steps that need to be completed for this task.
- [x] Add interface for inventories (12 Aug)
- [x] Add inventory class with basic functionality (13 Aug)
- [x] Test inventory class (13 Aug)
- [x] Merge into 3-inventory-and-items (16 Aug)
- [x] Write wiki (16 Aug)
- [ ] Merge into main (27 Aug)

# **Dependencies**
- [x] Item Abstract Class needs to be generated for storage of items (#18 )
 
**Completion Deadline:** Aug. 16

# **Documentation**
- [Inventory Wiki Page](https://github.com/UQcsse3200/2024-studio-2/wiki/Inventory)
 
# **Member**
- Name (Github Username) `Discord Username`
- Mittun (@mitsudz ) `mitsudz`

### LINKED ISSUE #24: Potion Abstract Class
# **Description**
**Task:** [Create Potion Abstract Class](#3).
**Feature:** [Items and Inventory](#2)

An abstract class to support various different potions (e.g, Health Potions, Attack Potions, Defence Potions). 
 
## **Example (optional)**
- **Animal Benefits**: Players can use potions to increase the health stats, attack stats and defence stats of the animal for duration of the battle. 

# **Milestones**
List of steps that need to be completed for this task.
- [x] Create Class Wireframe (Aug. 13)
- [x] Create Potion Sprite (Aug. 18)
 
**Completion Deadline:** Aug. 18

# **Documentation**
- TO BE COMPLETED
 
# **Member**
- Name (GitHub Username) (Slack Handle)
- e.g. Lewis (@lhick2108) (Lewis)

### LINKED ISSUE #27: Create Consumable Abstract class
# **Description**
**Task:** [Create Cinsumable Abstract Class](#3).
**Feature:** [#3](#2)

An abstract class to support various item types (e.g. items with a finite number of uses)

## **Example**
- Food items for animal hunger during a match and cannot perform properly until it is fed so the food item will be available for them to consume 
- Potions to improve animal stats

# **Dependencies**
- [x] Items Abstract class (#18 )
 
# **Milestones**
List of steps that need to be completed for this task.
- [x] Create Consumable Class (Aug. 14)
- [x] Create tests for Consumable Class (Aug. 14)
- [x] Merge into 3-inventory-and-items via pull request (Team 3 main branch) (Aug 14)
- [ ]  Add documentation for Consumable class in wiki (Aug. 16)

 
**Completion Deadline:** Aug. 16

 
# **Documentation**
- [Consumable Item Wiki](https://github.com/UQcsse3200/2024-studio-2/wiki/Consumable-Item) ** NOT YET COMPLETED **
 
# **Member**
- Name (@PHONGNGYUEN1 ) (beannie)

### LINKED ISSUE #28: Create Food Abstract class and Render Food Items
# **Description**
**Task:** [Create Food Abstract class](#3).
**Feature:** [#3 ](#2)

An abstract class to support different food types (e.g. sea, land, sky based food types) to feed to animals mid or outside of combat. This should also involve rendering 
 
## **Example (optional)**
- Animal may feel hungry during combat and requires food items to continue to perform 

# **Dependencies**
- [x] #18 
- [x] #27 

# **Milestones**
List of steps that need to be completed for this task.
- [x] Create Item class (Aug. 11)
- [x] Create Consumables  (Aug. 12)
- [x] Create Food abstract class (Aug. 12)
- [x] Create Multiple food items (Aug. 23) 
- [x] Create food assets(Aug. 20) 
 
**Completion Deadline:** Aug. 26

 
# **Documentation**
- [Food Wiki Page](https://github.com/UQcsse3200/2024-studio-2/wiki/Food) (NOT YET UPDATED TO INCLUDE RENDER INFORMATION)
 
# **Member**
- Name (@PHONGNGYUEN1) (beannie) (Nicole)

### LINKED ISSUE #30: Weapons
# **Description**
**Task:** Create class for infinite use items (#3).
**Feature:** Items and Inventory (#2)

Create a class for infinite use items (InfiniteUse), a subclass of AbstractItem. Infinite use items can be used multiple times without depletion. They include weapons and story key items, which are essential for gameplay progression.

# **Dependencies**
- [ ] Completion of AbstractItem class (#19 )
- [ ] Create specific story key elements (Discuss with quest and progression team)
 
# **Milestones**
- [ ] Initial design (Aug. 18)
- [ ] Unit tests implemented and initial documentation on wiki (Aug. 20)
- [ ] Class implemented and tested (Aug. 24)
- [ ] Documentation created and updated on wiki (Aug. 27)
 
**Completion Deadline:** Aug. 29

# **Documentation**
TBC
 
# **Member**
- Senuri Panadura Arachchige (s-shalomi)

### LINKED ISSUE #62: Inventory UI
# **Description**
**Task:** [Inventory Render Component](#3 ).
**Feature:** [Items and Inventory](#3 )

A class to support various types of inventory's (e.g. combat inventory, general inventory etc.). Should implement abstract interface for handling using, adding and removing from inventory. 
 
# **Milestones**
List of steps that need to be completed for this task.
- [x] Create an Inventory Component for the Player Entity to hold (18 Aug)
- [ ] Test Inventory Component (18 Aug)
- [x] Create Inventory Render Component and implement textures (19 Aug)
- [x] Ensure Inventory is renderable on screen via button (e.g. 'u') press (19 Aug)
- [ ] Create inventory button on screen  (20 Aug)
- [ ] Test Render Component for mouse click on Inventory button (20 Aug)
- [ ] Write wiki (21 Aug)
 
**Completion Deadline:** Aug. 21

# **Documentation**
- [Inventory Wiki Page](https://github.com/UQcsse3200/2024-studio-2/wiki/Inventory) (NOT YET UPDATED TO INCLUDE RENDER INFORMATION)
 
# **Member**
- Name (Github Username) `Discord Username`
- Mittun (@mitsudz ) `mitsudz`

### LINKED ISSUE #94: Abstract Item V2
# **Description**
**Task:** Implement new features to abstract item.
**Feature:** Inventory and Items (#3)

Abstract class will be updated so that inventory items can be rendered. Abstract item class will also have an item description which will show items abilities and stats. Item concrete sub-classes will also have to be modified so that the inherit the updated abstract item functionality. 

# **Milestones**
List of steps that need to be completed for this task.
- [x] get texture in abstract item (Aug. 22)
- [x] get description in abstract item (Aug. 22)
- [x] update concrete sub-classes (Aug. 23)
- [x] update all test and java docs (Aug. 24)
- [ ] Merge into main (27 Aug)
 
**Completion Deadline:** Aug. 28

# **Documentation**
- [Items Wiki Page](https://github.com/UQcsse3200/2024-studio-2/wiki/Items)
 
# **Member**
- Tom Preston (tompreston1) 
- Phonh Nguyen (PHONGNGUYEN1)

### LINKED ISSUE #95: Inventory Management
# **Description**
**Task:** Managing and sorting inventory (#3).
**Feature:** Inventory and Items (#2)

Update inventory class to enable categorising and sorting inventory by item name/code.

# **Dependencies**
- [x] AbstractItem (#18)
- [x] Concrete Item classes 
 
# **Milestones**
- [x] Update AbstractItem class with a category and description field and relevant accessor and mutator methods
- [x] Implement sorting functionality, based on item name, item code, quantity
- [ ] Implement a categorization functionality
- [ ] Implement search functionality, based on item name, item code, 
- [ ] Merge into main

**Completion Deadline:** Aug. 25
 
# **Documentation**
- [Inventory Wiki Page](https://github.com/UQcsse3200/2024-studio-2/wiki/Items)
 
# **Member**
- Name (GitHub Username) (Slack Handle)
- Senuri Panadura Arachchige (s-shalomi)

### LINKED ISSUE #105: Place Items in Map
# **Description**
**Task:** [Place Items in Map](#3).
**Feature:** [#3 ]

Statically place rendered items into the map using sprite and make event handling to pick up items from the map and add to the inventory.
 
## **Example**
- **Picking up**: The player encounters a candy near its proximity and decides to collect it to use for later. This candy will then be stored in the inventory
 
# **Milestones**
List of steps that need to be completed for this task.
- [x] Create food items and Potions items(Aug. 22)
- [x] Update map (Aug. 26)
- [x] Allow items to be picked up when in proximity of player by button press
- [x]  Display those items in the inventory
- [ ] Merge into main

**Completion Deadline:** Aug. 26
 
# **Documentation**
- [Items Wiki Page](https://github.com/UQcsse3200/2024-studio-2/wiki/Items)
- [Inventory Wiki Page](https://github.com/UQcsse3200/2024-studio-2/wiki/Inventory)

 
# **Member**
- (PHONGNGUYEN1) (Phong Nguyen)
- (lhick2108) (Lewis)

### PR DESCRIPTION
**Pull request by mitsudz**
Adding Inventory and Items to Main.

closes #3 
and thereby:
closes #18 
closes #19 
closes #24 
closes #27 
closes #28 
closes #30 
closes #62 
closes #94 
closes #95 
closes #105

### CODE REVIEW

**Comment by lhick2108** [comment]
Is it possible for you to pull in my documentation #105 place items in map so that there is clear documentation of ItemFactory and ItemProximityTask?

**Review by BaileyS03 — CHANGES_REQUESTED** [review]
Hi, the code looks pretty much fine. Do you have any unit testing plans // other testing plans? I couldnt see many sorry!

**Comment by PHONGNGYUEN1** [comment]
@BaileyS03 Yes we do, not quite sure why it doesn't show all though in git hub but we do have lots of tests written already for each file we created

**Review by PHONGNGYUEN1 — APPROVED** [review]
Code is working accordingly, ready to merge

**Comment by BaileyS03** [comment]
@PHONGNGYUEN1 
image

**Review by BaileyS03 — APPROVED** [review]
Thanks for fixing those tests

File: source/core/assets/images/foodtextures/download.png
Comment (jakingy) [inline source/core/assets/images/foodtextures/download.png:1]: duplicate file to appleItem.png with vague name, can remove

File: source/core/assets/images/potiontexture/Health potion.png
Comment (jakingy) [inline source/core/assets/images/potiontexture/Health potion.png:1]: more duplicate images, is there a reason for this?

**Review by jakingy — APPROVED** [review]
A few duplicate images, but code style is good and there are sufficient tests.

**Pull request merged**